src/main/java/com/redhat/rhjmc/containerjfr/net/web/WebServer.java

```diff
@@ -223,6 +225,53 @@ private void endWithJsonKeyValue(String key, String value, HttpServerResponse re
         response.end(String.format("{\"%s\":\"%s\"}", key, value));
     }
 
+    private HttpServerResponse writeInputStreamSync(InputStream inputStream, HttpServerResponse response) throws IOException {
+        // blocking function, must be called from a blocking handler
+        byte[] buff = new byte[WRITE_BUFFER_SIZE];
+        Buffer chunk = Buffer.buffer();
+
+        CompletableFuture<Void> future = new CompletableFuture<>();
+
+        (new Runnable() {
+            @Override
+            public void run() {
+                int n;
+                try {
+                    n = inputStream.read(buff);
+                } catch (IOException e) {
+                    future.completeExceptionally(e);
+                    return;
+                }
+
+                if (n == -1) {
+                    future.complete(null);
+                    return;
+                }
+
+                chunk.setBytes(0, buff, 0, n);
+                response.write(chunk.slice(0, n), (res) -> {
+                    if (res.failed()) {
+                        future.completeExceptionally(res.cause());
+                        return;
+                    }
+                    run(); // recursive call on this runnable itself
+                });
+            }
+        }).run();
+
+        try {
+            future.join();
```

Comment:
This looks like it's going to do what we talked about, but I can imagine this easily becoming confusing in the future when someone goes back to look at this implementation. In particular, I think it should be pointed out that the first execution of the runnable is done by whatever thread is handling the request, while the subsequent invocations are actually going to be run by some other thread specified by the HttpServerResponse and however it handles its internal write queue. This actually seems a bit off to begin with - would it make sense to submit the initial Runnable to an ExecutorService, and then in the `write` completion handler, submit the subsequent re-run to the same executor? ie the only things done by the vertx write queue handler are `future.completeExceptionally` and `executor.submit(this)`, with all of the buffer/chunk management and handing off the chunk data to the write queue done by the same ExecutorService thread. You could also use `shutdown/awaitTermination` on the ExecutorService to replace your explicitly managed `future`.

Reply:
Actually maybe you still would need the `future`, since I don't know how else you would nicely signal exceptional statuses.

Reply:
You're right. Sorry, I was lacking some consideration there, and thanks for pointing out using `ExecutorService`. I didn't eliminate the usage of `future` as `.awaitTermination()` requires a timeout parameter, and such timeout can not be easily determined since the recording size and network condition is unknown. (And having a timeout makes little senses if we are streaming a file.)

Reply:
Fair enough. The timeout could have simply been arbitrarily large, like Long.MAX_VALUE * TimeUnit.DAYS (or maybe be a little more sensible), but the implementation you have here now looks good.